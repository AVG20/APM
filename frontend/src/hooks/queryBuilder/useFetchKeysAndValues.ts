/* eslint-disable sonarjs/cognitive-complexity */
import { getAttributesValues } from 'api/queryBuilder/getAttributesValues';
import { DEBOUNCE_DELAY } from 'constants/queryBuilderFilterConfig';
import {
	getRemovePrefixFromKey,
	getTagToken,
	isInNInOperator,
} from 'container/QueryBuilder/filters/QueryBuilderSearch/utils';
import useDebounceValue from 'hooks/useDebounce';
import { isEqual, uniqWith } from 'lodash-es';
import { useCallback, useMemo, useRef, useState } from 'react';
import { useDebounce } from 'react-use';
import {
	BaseAutocompleteData,
	DataTypes,
} from 'types/api/queryBuilder/queryAutocompleteResponse';
import { IBuilderQuery } from 'types/api/queryBuilder/queryBuilderData';
import { DataSource } from 'types/common/queryBuilder';

import { useGetAggregateKeys } from './useGetAggregateKeys';

type IuseFetchKeysAndValues = {
	keys: BaseAutocompleteData[];
	results: string[];
	isFetching: boolean;
	sourceKeys: BaseAutocompleteData[];
	handleRemoveSourceKey: (newSourceKey: string) => void;
};

/**
 * Custom hook to fetch attribute keys and values from an API
 * @param searchValue - the search query value
 * @param query - an object containing data for the query
 * @returns an object containing the fetched attribute keys, results, and the status of the fetch
 */

export const useFetchKeysAndValues = (
	searchValue: string,
	query: IBuilderQuery,
	searchKey: string,
): IuseFetchKeysAndValues => {
	const [sourceKeys, setSourceKeys] = useState<BaseAutocompleteData[]>([]);
	const [results, setResults] = useState<string[]>([]);

	const memoizedSearchParams = useMemo(
		() => [
			searchKey,
			query.dataSource,
			query.aggregateOperator,
			query.aggregateAttribute.key,
		],
		[
			searchKey,
			query.dataSource,
			query.aggregateOperator,
			query.aggregateAttribute.key,
		],
	);

	const searchParams = useDebounceValue(memoizedSearchParams, DEBOUNCE_DELAY);

	const isQueryEnabled =
		query.dataSource === DataSource.METRICS
			? !!query.aggregateOperator &&
			  !!query.dataSource &&
			  !!query.aggregateAttribute.dataType
			: true;

	const { isFetching } = useGetAggregateKeys(
		{
			searchText: searchKey,
			dataSource: query.dataSource,
			aggregateOperator: query.aggregateOperator,
			aggregateAttribute: query.aggregateAttribute.key,
			tagType: query.aggregateAttribute.type ?? null,
		},
		{
			queryKey: [searchParams],
			enabled: isQueryEnabled,
			onSuccess: (data) => {
				if (data && data.payload) {
					setSourceKeys((prevState) =>
						uniqWith([...(data.payload.attributeKeys ?? []), ...prevState], isEqual),
					);
				}
			},
		},
	);

	/**
	 * Fetches the options to be displayed based on the selected value
	 * @param value - the selected value
	 * @param query - an object containing data for the query
	 */
	const handleFetchOption = async (
		value: string,
		query: IBuilderQuery,
		keys: BaseAutocompleteData[],
	): Promise<void> => {
		if (!value) {
			return;
		}
		const { tagKey, tagOperator, tagValue } = getTagToken(value);
		const filterAttributeKey = keys.find(
			(item) => item.key === getRemovePrefixFromKey(tagKey),
		);
		setResults([]);

		if (!tagKey || !tagOperator) {
			return;
		}

		const { payload } = await getAttributesValues({
			aggregateOperator: query.aggregateOperator,
			dataSource: query.dataSource,
			aggregateAttribute: query.aggregateAttribute.key,
			attributeKey: filterAttributeKey?.key ?? tagKey,
			filterAttributeKeyDataType: filterAttributeKey?.dataType ?? DataTypes.EMPTY,
			tagType: filterAttributeKey?.type ?? '',
			searchText: isInNInOperator(tagOperator)
				? tagValue[tagValue.length - 1]?.toString() ?? '' // last element of tagvalue will be always user search value
				: tagValue?.toString() ?? '',
		});

		if (payload) {
			const values = Object.values(payload).find((el) => !!el) || [];
			setResults(values);
		}
	};

	const handleRemoveSourceKey = useCallback((sourceKey: string) => {
		setSourceKeys((prevState) =>
			prevState.filter((item) => item.key !== sourceKey),
		);
	}, []);

	// creates a ref to the fetch function so that it doesn't change on every render
	const clearFetcher = useRef(handleFetchOption).current;

	// debounces the fetch function to avoid excessive API calls
	useDebounce(() => clearFetcher(searchValue, query, sourceKeys), 750, [
		clearFetcher,
		searchValue,
		query,
		searchKey,
	]);

	return {
		keys: sourceKeys,
		results,
		isFetching,
		sourceKeys,
		handleRemoveSourceKey,
	};
};
