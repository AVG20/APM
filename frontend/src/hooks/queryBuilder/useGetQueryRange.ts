import { REACT_QUERY_KEY } from 'constants/reactQueryKeys';
import {
	GetMetricQueryRange,
	GetQueryResultsProps,
} from 'lib/dashboard/getQueryResults';
import { useMemo } from 'react';
import { useQuery, UseQueryOptions, UseQueryResult } from 'react-query';
import { SuccessResponse } from 'types/api';
import { MetricRangePayloadV3 } from 'types/api/metrics/getQueryRange';

type UseGetQueryRange = (
	requestData: GetQueryResultsProps,
	options?: UseQueryOptions<SuccessResponse<MetricRangePayloadV3>, Error>,
) => UseQueryResult<SuccessResponse<MetricRangePayloadV3>, Error>;

export const useGetQueryRange: UseGetQueryRange = (requestData, options) => {
	const queryKey = useMemo(() => {
		if (options?.queryKey) {
			return [...options.queryKey];
		}
		return [REACT_QUERY_KEY.GET_QUERY_RANGE, requestData];
	}, [options?.queryKey, requestData]);

	return useQuery<SuccessResponse<MetricRangePayloadV3>, Error>({
		queryFn: async () => GetMetricQueryRange(requestData),
		...options,
		queryKey,
	});
};
