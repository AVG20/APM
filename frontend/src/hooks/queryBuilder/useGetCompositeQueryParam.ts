import { queryParamNamesMap } from 'constants/queryBuilderQueryNames';
import useUrlQuery from 'hooks/useUrlQuery';
import { useMemo } from 'react';
import { Query } from 'types/api/queryBuilder/queryBuilderData';

export const useGetCompositeQueryParam = (): Query | null => {
	const urlQuery = useUrlQuery();

	return useMemo(() => {
		const compositeQuery = urlQuery.get(queryParamNamesMap.compositeQuery);

		return compositeQuery ? JSON.parse(compositeQuery) : null;
	}, [urlQuery]);
};
