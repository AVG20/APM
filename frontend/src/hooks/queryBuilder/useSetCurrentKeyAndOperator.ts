import {
	getRemovePrefixFromKey,
	getTagToken,
} from 'container/QueryBuilder/filters/QueryBuilderSearch/utils';
import { useRef } from 'react';
import { BaseAutocompleteData } from 'types/api/queryBuilder/queryAutocompleteResponse';

type ICurrentKeyAndOperator = [string, string, string[]];

export const useSetCurrentKeyAndOperator = (
	value: string,
	keys: BaseAutocompleteData[],
): ICurrentKeyAndOperator => {
	const keyRef = useRef<string>('');
	const operatorRef = useRef<string>('');
	let result: string[] = [];

	const { tagKey, tagOperator, tagValue } = getTagToken(value);

	const isSuggestKey = keys?.some(
		(el) => el?.key === getRemovePrefixFromKey(tagKey),
	);
	if (isSuggestKey || keys.length === 0) {
		keyRef.current = tagKey || '';
		operatorRef.current = tagOperator || '';
		result = tagValue || [];
	}

	return [keyRef.current, operatorRef.current, result];
};
