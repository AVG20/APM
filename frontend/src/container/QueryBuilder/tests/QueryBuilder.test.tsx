import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import { I18nextProvider } from 'react-i18next';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import i18n from 'ReactI18';
import store from 'store';
import { DataSource } from 'types/common/queryBuilder';

import { QueryBuilder } from '../QueryBuilder';

describe('QueryBuilder', () => {
	it('renders the QueryBuilder component', () => {
		render(
			<QueryBuilder
				config={{
					queryVariant: 'static',
					initialDataSource: DataSource.METRICS,
				}}
				panelType="TIME_SERIES"
			/>,
		);
	});

	it('adds a new query when the "Query" button is clicked', () => {
		const { getByText } = render(
			<QueryBuilder
				config={{
					queryVariant: 'static',
					initialDataSource: DataSource.METRICS,
				}}
				panelType="TIME_SERIES"
			/>,
		);

		fireEvent.click(getByText('Query'));

		expect(getByText('Query')).toBeInTheDocument();
	});

	it('adds a new formula when the "Formula" button is clicked', () => {
		const { getByText } = render(
			<QueryBuilder
				config={{
					queryVariant: 'static',
					initialDataSource: DataSource.METRICS,
				}}
				panelType="TIME_SERIES"
			/>,
		);

		fireEvent.click(getByText('Formula'));

		expect(getByText('Formula')).toBeInTheDocument();
	});
});

describe('QueryBuilder', () => {
	it('should render QueryBuilder', () => {
		const { asFragment } = render(
			<MemoryRouter>
				<Provider store={store}>
					<I18nextProvider i18n={i18n}>
						<QueryBuilder config={undefined} panelType="TIME_SERIES" />
					</I18nextProvider>
				</Provider>
			</MemoryRouter>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
