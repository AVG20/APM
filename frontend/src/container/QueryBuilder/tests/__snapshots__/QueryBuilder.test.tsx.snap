// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query Builder it should render QueryBuilder 1`] = `
<DocumentFragment>
  <div
    class="ant-row ant-row-start css-dev-only-do-not-override-1i536d8"
    style="margin-top: -10px; margin-bottom: -10px;"
  >
    <div
      class="ant-col ant-col-24 css-dev-only-do-not-override-1i536d8"
      style="padding-top: 10px; padding-bottom: 10px;"
    >
      <div
        class="ant-row css-dev-only-do-not-override-1i536d8"
        style="margin-top: -25px; margin-bottom: -25px;"
      />
    </div>
    <div
      class="ant-row css-dev-only-do-not-override-1i536d8"
      style="margin-left: -10px; margin-right: -10px;"
    >
      <div
        class="ant-col css-dev-only-do-not-override-1i536d8"
        style="padding-left: 10px; padding-right: 10px;"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1i536d8 ant-btn-primary"
          type="button"
        >
          <span
            aria-label="plus"
            class="anticon anticon-plus"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </span>
          <span>
            Query
          </span>
        </button>
      </div>
      <div
        class="ant-col css-dev-only-do-not-override-1i536d8"
        style="padding-left: 10px; padding-right: 10px;"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-1i536d8 ant-btn-primary"
          type="button"
        >
          <span
            aria-label="plus"
            class="anticon anticon-plus"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="plus"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <defs>
                <style />
              </defs>
              <path
                d="M482 152h60q8 0 8 8v704q0 8-8 8h-60q-8 0-8-8V160q0-8 8-8z"
              />
              <path
                d="M176 474h672q8 0 8 8v60q0 8-8 8H176q-8 0-8-8v-60q0-8 8-8z"
              />
            </svg>
          </span>
          <span>
            Formula
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
